File: python/saddle-points/saddle_points.py
def saddle_points(matrix):
    if len(matrix) > 1:
        prev = matrix[0]
        for m in matrix[1::]:
            if len(m) != len(prev): raise ValueError("irregular matrix")
            prev = m

    possible = []
    for x, row in enumerate(matrix):
        max_row_height = max(row)
        for y, cell_height in enumerate([*row]):
            if cell_height == max_row_height:
                column = [r[y] for r in matrix]
                min_col_height = min(column)
                if cell_height <= min_col_height:
                    possible.append({"row": x+1, "column": y+1})

    return possible

File: ruby/saddle-points/saddle_points.rb
class Grid
  def self.saddle_points(rows)
    cols = rows.transpose
    out = []
    rows.each_with_index do |row, y|
      row.each_with_index do |height, x|
        out.append({
          "row"    => y + 1,
          "column" => x + 1,
        }) if height == row.max and height == cols[x].min
      end
    end
    return out
  end
end
